libraries/c_sdk/standard/https/src/iot_https_client.c

```diff
@@ -0,0 +1,3326 @@
+    {
+        /* For a HEAD method, there is no body expected in the response, so we return 1 to skip body parsing. */
+        if( ( pHttpsResponse->method == IOT_HTTPS_METHOD_HEAD ) )
+        {
+            retVal = STOP_PARSING;
+
+            /* Since the message is considered complete now for a HEAD response, then we set the parser state
+             * to the completed state. */
+            pHttpsResponse->parserState = PARSER_STATE_BODY_COMPLETE;
+        }
+
+        /* If this is NOT a HEAD method and there is body configured, but the server does not send a body in the
+         * response, then the body buffer will be filled with the zeros from rest of the header buffer. http-parser
+         * will invoke the on_body callback and consider the zeros following the headers as body. */
+
+        /* If it was configured in a synchronous response to ignore the HTTPS response body then also stop the body
+         * parsing. There could be body in the message, despite the body configured to be ignore, so the parser state
+         * stays as is so that the network socket can be flushed of the body if needed. */
+        if( ( pHttpsResponse->isAsync == false ) && ( pHttpsResponse->pBody == NULL ) )
+        {
+            retVal = STOP_PARSING;
+        }
+    }
+
+    return retVal;
+}
+
+/*-----------------------------------------------------------*/
+
+static int _httpParserOnBodyCallback( http_parser * pHttpParser,
+                                      const char * pLoc,
+                                      size_t length )
+{
+    IotLogDebug( "Parser: Reached the HTTPS message body. It is of length: %d", length );
+
+    _httpsResponse_t * pHttpsResponse = ( _httpsResponse_t * ) ( pHttpParser->data );
+    pHttpsResponse->parserState = PARSER_STATE_IN_BODY;
+
+    if( ( pHttpsResponse->bufferProcessingState == PROCESSING_STATE_FILLING_HEADER_BUFFER ) && ( pHttpsResponse->isAsync ) )
+    {
+        /* For an asynchronous response, the buffer to store the body will be available after the headers
+         * are read first. We may receive part of the body in the header buffer. We will want to leave this here
+         * and copy it over when the body buffer is available in the _readReadyCallback().
+         */
+        if( pHttpsResponse->pBodyInHeaderBuf == NULL )
+        {
+            pHttpsResponse->pBodyInHeaderBuf = ( uint8_t * ) ( pLoc );
+            pHttpsResponse->pBodyCurInHeaderBuf = pHttpsResponse->pBodyInHeaderBuf;
+        }
+
+        /* If there is a chunk encoded body in the header buffer, we will want to overwrite the chunk headers with the
+         * actual body. This is so that when the application calls IotHttpsClient_ReadResponseBody(), in the
+         * readReadyCallback(), we can pass the body into the body buffer provided right away. */
+        if( pHttpsResponse->pBodyCurInHeaderBuf != ( uint8_t * ) pLoc )
+        {
+            memcpy( pHttpsResponse->pBodyCurInHeaderBuf, pLoc, length );
```

Comment:
This block would happen when we just got the remainder, or all, of the head, plus a chunk of the body? If so, could that be explained in more detail in the comments?